src/TorchSharp/JIT/Type/Type.cs

```diff
@@ -1,108 +1,126 @@
-//        /// </summary>
-//        protected void Dispose(bool disposing)
-//        {
-//            if (disposing)
-//            {
-//                handle.Dispose();
-//                handle.SetHandleAsInvalid();
-//            }
-//        }
-
-//        [DllImport("LibTorchSharp")]
-//        private static extern sbyte THSJIT_typeKind(HType handle);
-
-//        internal TypeKind Kind
-//        {
-//            get { return (TypeKind)THSJIT_typeKind(handle); }
-//        }
-
-//        [DllImport("LibTorchSharp")]
-//        private static extern IntPtr THSJIT_typeCast(HType module);
-
-//        internal TensorType AsTensorType()
-//        {
-//            return new TensorType(THSJIT_typeCast(handle));
-//        }
-
-//        internal DynamicType AsDynamicType()
-//        {
-//            return new DynamicType(THSJIT_typeCast(handle));
-//        }
-
-//        internal enum TypeKind : sbyte
-//        {
-//            TensorType = 0,
-//            DimensionedTensorType = 1
-//        }
-//    }
-//}
+// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+using System;
+using System.Collections;
+using System.Runtime.InteropServices;
+
+namespace TorchSharp
+{
+    public static partial class torch
+    {
+        public static partial class jit
+        {
+
+            public class Type : IDisposable
+            {
+                /// <summary>
+                ///    Class wrapping PyTorch's type object reference.
+                /// </summary>
+                internal sealed class HType : SafeHandle
+                {
+                    public HType(IntPtr preexistingHandle, bool ownsHandle, TypeKind kind) : base(IntPtr.Zero, ownsHandle)
+                    {
+                        SetHandle(preexistingHandle);
+                        this.kind = kind;
+                    }
+
+                    public override bool IsInvalid => handle == IntPtr.Zero;
+
+                    // This is just for marshalling
+                    internal HType() : base(IntPtr.Zero, true)
+                    {
+                    }
+
+                    [DllImport("LibTorchSharp")]
+                    private static extern void THSJIT_Type_dispose(HType handle);
+
+                    [DllImport("LibTorchSharp")]
+                    private static extern void THSJIT_TensorType_dispose(HType handle);
+
+                    protected override bool ReleaseHandle()
+                    {
+                        switch (kind) {
+                        case TypeKind.TensorType:
+                            THSJIT_TensorType_dispose(this);
+                            break;
+                        default:
+                            THSJIT_Type_dispose(this);
+                            break;
+                        }
+                        return true;
+                    }
+
+                    protected override void Dispose(bool disposing)
```

Comment:
>Dispose [](http://example.com/codeflow?start=44&length=7)

I think you don't need to override this method. The SafeHandle.Dispose(bool) handle the internal disposing state of the object and calls ReleaseHandle as needed. 

[https://source.dot.net/#System.Private.CoreLib/SafeHandle.cs,98](https://source.dot.net/#System.Private.CoreLib/SafeHandle.cs,98)